src/stan/services/util/initialize.hpp

```diff
@@ -13,202 +13,211 @@
+                              &msg);
+      }
+    } catch (std::domain_error& e) {
+      if (msg.str().length() > 0)
+        logger.info(msg);
+      logger.info("Rejecting initial value:");
+      logger.info("  Error evaluating the log probability"
+                  " at the initial value.");
+      logger.info(e.what());
+      continue;
+    } catch (std::exception& e) {
+      if (msg.str().length() > 0)
+        logger.info(msg);
+      logger.info("Unrecoverable error evaluating the log probability"
+                  " at the initial value.");
+      logger.info(e.what());
+      throw;
+    }
 
-            std::stringstream msg2;
-            msg2 << "1000 transitions using 10 leapfrog steps"
-                 << " per transition would take"
-                 << " " << 1e4 * deltaT << " seconds.";
-            logger.info(msg2);
+    msg.str("");
+    double log_prob(0);
+    try {
+      log_prob = model.template log_prob<false, true>
+                 (unconstrained, disc_vector, &msg);
+      if (msg.str().length() > 0)
+        logger.info(msg);
+    } catch (std::domain_error& e) {
+      if (msg.str().length() > 0)
+        logger.info(msg);
+      logger.info("Rejecting initial value:");
+      logger.info("  Error evaluating the log probability"
+                  " at the initial value.");
+      logger.info(e.what());
+      continue;
+    } catch (std::exception& e) {
+      if (msg.str().length() > 0)
+        logger.info(msg);
+      logger.info("Unrecoverable error evaluating the log probability"
+                  " at the initial value.");
+      logger.info(e.what());
+      throw;
+    }
+    if (!boost::math::isfinite(log_prob)) {
+      logger.info("Rejecting initial value:");
+      logger.info("  Log probability evaluates to log(0),"
+                  " i.e. negative infinity.");
+      logger.info("  Stan can't start sampling from this"
+                  " initial value.");
+      continue;
+    }
+    std::stringstream log_prob_msg;
+    std::vector<double> gradient;
+    bool gradient_ok = true;
+    clock_t start_check = clock();
+    try {
+      log_prob = stan::model::log_prob_grad<true, true>
+                 (model, unconstrained, disc_vector,
+                  gradient, &log_prob_msg);
+    } catch (const std::exception& e) {
+      if (log_prob_msg.str().length() > 0)
+        logger.info(log_prob_msg);
+      logger.info(e.what());
+      throw;
+    }
+    clock_t end_check = clock();
+    double deltaT = static_cast<double>(end_check - start_check)
+                    / CLOCKS_PER_SEC;
+    if (log_prob_msg.str().length() > 0)
+      logger.info(log_prob_msg);
 
-            logger.info("Adjust your expectations accordingly!");
-            logger.info("");
-            logger.info("");
-          }
-          if (gradient_ok)
-            break;
-        }
+    for (size_t i = 0; i < gradient.size(); ++i) {
+      if (gradient_ok && !boost::math::isfinite(gradient[i])) {
```

Comment:
This can only print once and it'd be much cleaner if there was an is_finite function.

```
template <class C>
bool is_finite(const C& gradient) {
  return boost::math::is_finite(sum(gradient));
}
```
or write it out in a loop if you're worried about efficiency.  Then this code is just:

```
if (!is_finite(gradient))
  do something about it
...
```

The major pain in pulling things out into functions is the extra doc and tests that some reviewers require.

Reply:
Definitely not worried about efficiency here. That looks simpler.

I don't find it to be a pain to split things into functions with doc and tests if it's reusable. Here, I think we have `sum` and `isfinite` in the math library, so we'll just use that.